src/BloomExe/web/ServerBase.cs

```diff
@@ -116,117 +74,122 @@ private bool StartWithExceptionHandling(out Exception error)
 		{
-			_listener = new HttpListener();
 			_listener.AuthenticationSchemes = AuthenticationSchemes.Anonymous;
 			_listener.Prefixes.Add(PathEndingInSlash);
 			_listener.Start();
-			//			_listener.BeginGetContext(new AsyncCallback(ListenerCallback), _listener);
 
-			_listenerThread = new Thread(HandleRequests);
 			_listenerThread.Start();
 
+			for (var i = 0; i < _workers.Length; i++)
+			{
+				_workers[i] = new Thread(WaitForRequests);
+				_workers[i].Start();
+			}
+
 			return GetIsAbleToUsePort();
 		}
 
-		private bool GetIsAbleToUsePort()
+		private void HandleRequests()
 		{
-			IsAbleToUsePort = false;
-			try
+			while (_listener.IsListening)
 			{
-				var x = new WebClientWithTimeout { Timeout = 3000 };
+				var context = _listener.BeginGetContext(ContextReady, null);
 
-				IsAbleToUsePort = ("OK" == x.DownloadString(PathEndingInSlash + "testconnection"));
-			}
-			catch (Exception e)
-			{
-				IsAbleToUsePort = false;
+				if (0 == WaitHandle.WaitAny(new[] { _stop, context.AsyncWaitHandle }))
+					return;
 			}
-			return IsAbleToUsePort;
 		}
 
-		private void HandleRequests()
+		private void ContextReady(IAsyncResult ar)
 		{
-			while (_listener.IsListening)
+			// this can happen when shutting down
+			if (!_listenerThread.IsAlive) return;
+
+			lock (_queue)
 			{
-				var context = _listener.BeginGetContext(new AsyncCallback(ListenerCallback), _listener);
-				if (0 == WaitHandle.WaitAny(new[] { _stop, context.AsyncWaitHandle }))
-					return;
+				_queue.Enqueue(_listener.EndGetContext(ar));
+				_ready.Set();
 			}
 		}
 
 		/// <summary>
-		/// TODO: Note: doing this at runtime isn't as good as doing it in the installer, because we have no way of
-		/// removing these entries on uninstall (but the installer does).
+		/// The worker threads run this function
 		/// </summary>
-		private static void AddUrlAccessControlEntry()
+		private void WaitForRequests()
 		{
-			MessageBox.Show(
-				string.Format("We need to do one more thing before Bloom is ready. Bloom needs temporary administrator privileges to set up part of its communication with the embedded web browser.{0}{0}After you click 'OK', you may be asked to authorize this step.", Environment.NewLine),
-				"Almost there!", MessageBoxButtons.OK);
+			// _ready: indicates that there are requests in the queue that should be processed.
+			// _stop:  indicates that the class is being disposed and the thread should terminate.
+			WaitHandle[] wait = { _ready, _stop };
 
-			var startInfo = new System.Diagnostics.ProcessStartInfo();
-			startInfo.UseShellExecute = true;
-			startInfo.Verb = "runas"; //makes it as for elevation to admin rights
-
-			if (Environment.OSVersion.Version.Major == 5 /*win xp*/)
-			{
-				startInfo.FileName = "httpcfg";
-				startInfo.Arguments = string.Format("set urlacl -u {0} /a \"D:(A;;GX;;;WD)\"", PathEndingInSlash);
-			}
-			else
+			// WaitHandle.WaitAny(wait) returns the array index of the handle that satisfied the wait. 
```

Comment:
This sort of gives you the information to figure things out, but it would be much more helpful to say something like, wait until a request is ready or the thread is being stopped. The WaitAny will return 0 (the index of _ready in the wait array) if a request is ready, and 1 when _stop is signaled, breaking us out of the loop.